[FIRRTL] Constant folding for Binary operations

prithayan opened the pull request:
Implemented the fold method for ``AddPrimOp``. 
1. If both the operands are (constant) ``IntegerAttr``, 
2. Get the ``APInt`` 
3. If signed get the sign-extended else zero-extended ``int64_t`` value from the ``APInt`` 
4. Add them and return a new ``IntegerAttr`` with incremented bitwidth.

Fix for : https://github.com/llvm/circt/issues/809

Will create separate PR for constant folding of other operators.

seldridge reviewed (commented):
This looks great. 

The other fold for `add` is `add(0, a) => a` or `add(a, 0) => a`. Does it make sense to also add this while you're here?

lattner commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -93,6 +93,27 @@ OpFoldResult ConstantOp::fold(ArrayRef<Attribute> operands) {
+  /// folding. Cannot use constFoldBinaryOp, since the width of the constant is
+  /// different from operands.
+  if (operands[0] && operands[1] &&
+      lhs().getType().cast<FIRRTLType>().getBitWidthOrSentinel() < 64) {
```
You shouldn't need to limit this to 64 bits (see below).

However, we have to be super careful about add operations that have operands/results that have unknown widths.  I'd recommend defining a general predicate that can be used by other folders so you can do something like this:

`  if (operands[0] && operands[1] && hasKnownWidths(this))`

hasKnownWidths would check that the operands and results of the Operation* are all integer types with known widths.

lattner commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -93,6 +93,27 @@ OpFoldResult ConstantOp::fold(ArrayRef<Attribute> operands) {
+  if (operands[0] && operands[1] &&
+      lhs().getType().cast<FIRRTLType>().getBitWidthOrSentinel() < 64) {
+    auto op1 = operands[0].dyn_cast<IntegerAttr>();
+    auto op2 = operands[1].dyn_cast<IntegerAttr>();
```
These two things can be `cast<>`'s.  We know that AddPrimOp takes firrtl integer operands, and this is the only thing that can come in here.

lattner commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -93,6 +93,27 @@ OpFoldResult ConstantOp::fold(ArrayRef<Attribute> operands) {
+    if (op1 && op2) {
+      APInt sumAPInt = (op1.getValue() + op2.getValue());
+      int64_t sumVal = sumAPInt.isSignBitSet() ? sumAPInt.getSExtValue()
+                                               : sumAPInt.getZExtValue();
```
Plz be careful to follow the implemention that LowerToRTL is using, this isn't correct.

In the case of add, we need to extend the LHS and RHS to the width of the destination (sign or zero extending depending on the type) and then do an add.  I'd recommend defining a helper function for this since the other constant folders will want to do a similar thing, e.g.:

`auto op1 = extendConstantToType(operands[0], getOperand(0), getType());` // returns an APInt

You need to pass in `getOperand(0)` because the sign/zero extension follows the sign of the operand, not the sign of the destination type.

Alternatively we could make sure that the attributes in FIRRTL::ConstantOp always have the correct sign.  If that were the case then we wouldn't need to pass in the type.  It looks like this could be true, but `verifyConstantOp` isn't verifying it, so it might not be true currently.

With this helper function then the body of this would just be:

```
auto op1 = extendConstantToType(operands[0], getOperand(0), getType());
auto op2 = extendConstantToType(operands[1], getOperand(1), getType());
return getIntAttr(op1+op2, getContext());
```

which also handles arbitrary width integers.

fabianschuiki commented:
We should also think about the case where the result is an unsized `UInt` or `SInt`. In that case we probably have to rely on the `APInt`'s own bit width to compute an appropriate bit width for the result `APInt`. Need some way to avoid duplicating all the result-bit-width computation logic that's already in the `FIRRTLOps.cpp` file. Maybe some separate function that computes the result bit width given input bit widths, that is then called from `getResultType` and in const folding whenever we need to operate directly on `APInt` bit widths for the sake of unsized `UInt`/`SInt`?

fabianschuiki commented on test/Dialect/FIRRTL/canonicalization.mlir:
```diff
@@ -514,4 +513,42 @@ firrtl.module @wire_port_prop1(%in_a: !firrtl.uint<9>, %out_b: !firrtl.flip<uint
+}
+
+// CHECK-LABEL: @add_cst_prop2
+// CHECK-NEXT:   %c127_si9 = firrtl.constant(127 : i9) : !firrtl.sint<9>
```
I think this constant should be `-1 : i9`, since you constant-fold `-6 + 5`. `-1` and `127` would have an identical 7 bit pattern (`0b111_1111`), but would differ for 9 bits (`0b1_1111_1111` vs `0b0_0111_1111`).

fabianschuiki commented on test/Dialect/FIRRTL/canonicalization.mlir:
```diff
@@ -514,4 +513,42 @@ firrtl.module @wire_port_prop1(%in_a: !firrtl.uint<9>, %out_b: !firrtl.flip<uint
+}
+
+// CHECK-LABEL: @add_cst_prop3
+// CHECK-NEXT:   %c-6_si4 = firrtl.constant(-6 : i4) : !firrtl.sint<4>
```
This seems to be related to the above. You would expect `-2 : i4` here due to folding of `-1 + -1`. `-6` coincides with zero-extending `-1 : i2` and `-1 : i3` to 4 bits, doing the addition, then interpreting the result as signed.

prithayan replied:
Thanks, fixed it now.

fabianschuiki commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -89,10 +89,52 @@ OpFoldResult ConstantOp::fold(ArrayRef<Attribute> operands) {
+    APInt sumAPInt =
+        extendConstantToWidth(
+            op1.getValue(),
+            getOperand(0).getType().cast<FIRRTLType>().isSignedInteger(),
```
Could it be that this `isSignedInteger()` always returns `false`, regardless of the type? That would explain the arithmetic mismatches you see further down in your test cases, because this would cause operands to be zero-extended all the time.

prithayan replied:
Thanks, that was the issue. I was wrongly using ``llvm::Type::isSignedInteger()``, updated it to ``FIRRTLType::IntType::isSigned()``

fabianschuiki reviewed (approved):
Neat :+1:

lattner commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -124,10 +130,72 @@ OpFoldResult ConstantOp::fold(ArrayRef<Attribute> operands) {
+/// Check if the results of \p op are of integer type and with known bitwidth.
+/// Can be used to determine if any fold is legal.
+bool hasKnownWidths(Operation *op) {
+  for (auto resTy : op->getResultTypes()) {
```
Please mark static, and please check operand types as well.

lattner commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -124,10 +130,72 @@ OpFoldResult ConstantOp::fold(ArrayRef<Attribute> operands) {
+}
+
+// Width extension of a constant int depends only on the sign of the operand.
+APInt extendConstantToWidth(APInt old, bool isSigned, unsigned newWidth) {
```
Plz mark static, also should this be used by `constFoldFIRRTLBinaryOp` above instead of duplicating the logic?

lattner commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -124,10 +130,72 @@ OpFoldResult ConstantOp::fold(ArrayRef<Attribute> operands) {
+  /// widths, then perform constant folding. Cannot use constFoldBinaryOp, since
+  /// the width of the constant is different from operands.
+  return constFoldFIRRTLBinaryOp<APInt>(
+      operands, getOperand(0).getType().cast<IntType>().isUnsigned(),
```
I think that each of these can be slightly simplified to use `getType().cast<IntType>().isUnsigned()`. instead of `getOperand(0).getType()...`, since the sign of the result type is always the sign of the operand type.

lattner reviewed (approved):
Looking good, thanks!

prithayan commented on include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:
Added the ``hasFolder=1`` to the ``BinaryPrimOp``, and dummy ``fold`` method for rest of the operations. Will handle the rest of the operations after Milestone-1.

prithayan commented:
Thanks for the detailed review. 

- Updated the ``constFoldFIRRTLBinaryOp`` to be more general for all the binary operations. 
- Set ``hasFolder=1`` for all the ``BinaryPrimOps``
- Added the fold method only for ``AddPrimOp, SubPrimOp, MulPrimOp``  and dummy folds for the rest. 
- Not handling the unknown bitwidth operand cases, which would be handled after width inference. 
- This is more conservative and handles few cases for Milestone-1. Will generalize it and handle rest of the cases after Milestone-1

lattner commented:
nice thanks!  One further simplification you could do: firrtl primitives are guaranteed to have a single result, so you could fold away one of the loops in hasKnownWidths
